misc/config_tools/launch_config/launch_script_template.sh

```diff
@@ -0,0 +1,166 @@
+#!/bin/bash
+#
+# Copyright (C) 2022 Intel Corporation.
+#
+# SPDX-License-Identifier: BSD-3-Clause
+#
+
+# Helper functions
+
+function probe_modules() {
+    modprobe pci_stub
+}
+
+function offline_cpus() {
+    # Each parameter of this function is considered the APIC ID (as is reported in /proc/cpuinfo, in decimal) of a CPU
+    # assigned to a post-launched RTVM.
+    for i in $*; do
+        processor_id=$(grep -B 15 "apicid.*: ${i}$" /proc/cpuinfo | grep "^processor" | head -n 1 | cut -d ' ' -f 2)
+        if [ -z ${processor_id} ]; then
+            continue
+        fi
+        cpu_path="/sys/devices/system/cpu/cpu${processor_id}"
+        if [ -f ${cpu_path}/online ]; then
+            online=`cat ${cpu_path}/online`
+            echo cpu${processor_id} online=${online} >> /dev/stderr
+            if [ "${online}" = "1" ] && [ "${processor_id}" != "0" ]; then
+                echo 0 > ${cpu_path}/online
+                online=`cat ${cpu_path}/online`
+                # during boot time, cpu hotplug may be disabled by pci_device_probe during a pci module insmod
+                while [ "${online}" = "1" ]; do
+                    sleep 1
+                    echo 0 > ${cpu_path}/online
+                    online=`cat ${cpu_path}/online`
+                done
+                echo ${processor_id} > /sys/devices/virtual/misc/acrn_hsm/remove_cpu
+            fi
+        fi
+    done
+}
+
+function unbind_device() {
+    physical_bdf=$1
+
+    vendor_id=$(cat /sys/bus/pci/devices/${physical_bdf}/vendor)
+    device_id=$(cat /sys/bus/pci/devices/${physical_bdf}/device)
+
+    echo $(printf "%04x %04x" ${vendor_id} ${device_id}) > /sys/bus/pci/drivers/pci-stub/new_id
+    echo ${physical_bdf} > /sys/bus/pci/devices/${physical_bdf}/driver/unbind
+    echo ${physical_bdf} > /sys/bus/pci/drivers/pci-stub/bind
+}
+
+function create_tap() {
+    # create a unique tap device for each VM
+    tap=$1
+    tap_exist=$(ip a | grep "$tap" | awk '{print $1}')
+    if [ "$tap_exist"x != "x" ]; then
+        echo "tap device existed, reuse $tap"
+    else
+        ip tuntap add dev $tap mode tap
+    fi
+
+    # if acrn-br0 exists, add VM's unique tap device under it
+    br_exist=$(ip a | grep acrn-br0 | awk '{print $1}')
+    if [ "$br_exist"x != "x" -a "$tap_exist"x = "x" ]; then
+        echo "acrn-br0 bridge aleady exists, adding new tap device to it..."
+        ip link set "$tap" master acrn-br0
+        ip link set dev "$tap" down
+        ip link set dev "$tap" up
+    fi
+}
+
+function mount_partition() {
+    partition=$1
+
+    tmpdir=`mktemp -d`
+    mount ${partition} ${tmpdir}
+    echo ${tmpdir}
```

Comment:
What is this? (I should know I guess but I'm blanking...)

Reply:
In today's launch XML configurations, users are allowed to specify the path of an image in the format `<partition>:<path in that partition>`. In that case the launch script will automatically mount that partition somewhere, using the rootfs image as specified in the path, and unmount the partition after the device model exits.

This helper function (and the `unmount_partition` as well) are the helper functions to implement to mount/unmount actions.

Reply:
Thanks for the explanation @junjiemao1 . And the `echo` command is there to make sure the right argument ends up in the launch script being generated, is that correct?

Reply:
Echoing the created directory is to allow the caller of this function to get the path so that it can be removed after the device model exits.

As bash function always returns exit codes, I have to use stdout as the channel for returning more info to the caller.